feat(route): add eventbrite

elibroftw opened the pull request:
<!-- 
If you have any difficulties in filling out this form, please refer to https://docs.rsshub.app/joinus/new-rss/submit-route
如果你在填写此表单时遇到任何困难，请参考 https://docs.rsshub.app/zh/joinus/new-rss/submit-route
-->

## Involved Issue / 该 PR 相关 Issue

N/A

## Example for the Proposed Route(s) / 路由地址示例

<!--
Please include route starts with /, with all required and optional parameters.
Fail to comply will result in your pull request being closed automatically.
请在 `routes` 区域填写以 / 开头的完整路由地址，否则你的 PR 将会被无条件关闭。
如果路由包含在文档中列出可以完全穷举的参数（例如分类），请依次全部列出。

```routes
/some/route
/some/other/route
/dont/use/this/or/modify/it
/use/the/fenced/code/block/below
```

If your changes are not related to route, please fill in `routes` section with `NOROUTE`. Fail to comply will result in your PR being closed.
如果你的 PR 与路由无关, 请在 `routes` 区域 填写 `NOROUTE`，而不是直接删除 `routes` 区域。否则你的 PR 将会被无条件关闭。
-->

```routes
/eventbrite/all-events/canada--toronto
```

## New RSS Route Checklist / 新 RSS 路由检查表
  
- [X] New Route / 新的路由
  - [X] Follows [Script Standard](https://docs.rsshub.app/joinus/advanced/script-standard) / 跟随 [路由规范](https://docs.rsshub.app/zh/joinus/advanced/script-standard)
- [ ] Anti-bot or rate limit / 反爬/频率限制
  - [ ] If yes, do your code reflect this sign? / 如果有, 是否有对应的措施?
- [ ] [Date and time](https://docs.rsshub.app/joinus/advanced/pub-date) / [日期和时间](https://docs.rsshub.app/zh/joinus/advanced/pub-date)
  - [X] Parsed / 可以解析
  - [ ] Correct time zone / 时区正确
- [ ] New package added / 添加了新的包
- [ ] `Puppeteer`

## Note / 说明

I don't know how to scope the radar to work on eventbrite.ca as well.

## Future Improvements

- Scrape max 5 or so pages (figure out how many new pages are added per day in NYC for example)
- Use destination/search API from the get-go instead of web scraping (requires the place ID which requires smart/regex webscraping)

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,71 @@
+    path: '/:eventType?/:region',
+    categories: ['programming'],
+    example: '/eventbrite/all-events/canada--toronto',
+    parameters: { evenType: 'category of events for filtering', region: 'Region or scope of events' },
```
```suggestion
    parameters: { eventType: 'category of events for filtering', region: 'Region or scope of events' },
```

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,71 @@
+import dayjs from 'dayjs';
+
+export const route: Route = {
+    path: '/:eventType?/:region',
```
Place the compulsory parameter before the optional parameter, otherwise the optional parameter becomes compulsory in order to fill in the compulsory parameter

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,71 @@
+        if (eventType === 'events') {
+            eventType = 'all-events';
+        }
+        const link = `https://eventbrite.com/d/${region}/${eventType}/`;
```
```suggestion
        const link = `https://www.eventbrite.com/d/${region}/${eventType}/`;
```

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,71 @@
+        const link = `https://eventbrite.com/d/${region}/${eventType}/`;
+        const response = await ofetch(link);
+        const $ = load(response);
+        const startTime = dayjs();
+        const items = $('div.search-results-panel-content section ul li')
+            .toArray()
+            .map((item, index): DataItem => {
+                // Add index parameter
+                const el = $(item);
+                const a = el.find('a.event-card-link').first();
+                const pElements = el.find('p');
+                return {
+                    title: a.attr('aria-label') || a.text(),
+                    link: a.attr('href')!,
+                    pubDate: startTime.add(index, 'second').toDate(), // Increment pubDate by 1 second for each item
+                    author: pElements.length > 1 ? $(pElements[1]).text() : '',
+                    category: ['event'],
+                };
+            });
+
+        const formattedEventType = capitalizeWords(eventType.replaceAll('-', ' '));
+        const formattedRegion = capitalizeWords(region.replaceAll('-', ' '));
```
Parse all `data-event-id` from `.event-card-link` and make an additional call to `/api/v3/destination/events/` which provides precise info regarding each event.

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,71 @@
+    },
+};
+
+function capitalizeWords(str: string): string {
+    return str
+        .split(' ')
+        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
+        .join(' ');
+}
```
You can use https://github.com/DIYgod/RSSHub/blob/0510d8391e08b2260787ca7955c88fd18ea4d478/lib/utils/common-utils.ts#L6

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,71 @@
+
+export const route: Route = {
+    path: '/:eventType?/:region',
+    categories: ['programming'],
```
```suggestion
    categories: ['other'],
```

elibroftw commented:
Addressed comments, however I still want this to support the .ca domain before merging as people in Canada like me will be using the .ca event brite not the .com

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,152 @@
+        }
+        const link = `https://www.eventbrite.com/d/${region}/${eventType}/`;
+        const response = await ofetch(link, {
+            headers: {
+                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:138.0) Gecko/20100101 Firefox/138.0',
+            },
```
Does the site only work on this fixed version of Firefox on Windows? If the answer is no, please remove it which RSSHub will use a randomised Chrome on macOS by default.

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,152 @@
+            if (eventData === undefined) {
+                const pElements = el.find('p');
+                return {
+                    title: `should be ${fallbackTitle}`,
+                    author: pElements.length > 1 ? `should be ${$(pElements[1]).text()}` : 'unknown author',
```
Do not use `should be` or `unknown author`. Leave it `undefined` if you don't know.

TonyRL commented on lib/routes/eventbrite/events.ts:
```diff
@@ -0,0 +1,152 @@
+                description: eventData.summary,
+                id: eventData.eventbrite_event_id,
+                content: {
+                    html: `${eventDate(eventData)}\n${getAuthor(eventData)}\n${getTicketPriceRange(eventData)}`,
```
```suggestion
                    html: `${eventDate(eventData)}<br>${getAuthor(eventData)}<br>${getTicketPriceRange(eventData)}`,
```
